### ISSUE TITLE
Determine how/whether napari will help download remote resources

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#3260]
## 🧰 Task
Just opening an issue prompted by  https://github.com/napari/napari/issues/316#issuecomment-906515843 to discuss how and whether napari will "wrap" remote resources in any way, to facilitate downloading and opening.

For example, in `scikit-image` the imread function [wraps the path in a `file_or_url_context`](https://github.com/scikit-image/scikit-image/blob/d85b06468e063f2b8aa6aa7e073e120294e8e2f8/skimage/io/_io.py#L47) that takes care of downloading the resource to a temporary file, then `urllib.request.urlopen()` to download the data, and only then passes the local data to the underlying plugin reader.  

Currently, we leave all management of remote data to plugins.  So things like `ome-zarr` or some of @manzt's plugins can do a great job loading from a particular remote store.  But everyone else, including our builtin magic_imread  is left to fend for themselves (see also where this came up in https://github.com/napari/napari/issues/3021#issuecomment-877796596).  so if the user has an `nd2` reader installed, they still can't load `https://..../somefile.nd2`, even though all the pieces are in place.

I don't have a particular proposal, but I think it'd be good to decide:

a) whether we intend to wrap certain protocols (maybe just `http`?  maybe others?) to provide some sort of wrapped temporary resource that will let plugins treat remote data as if it was on the local file system.
b) if so, decide which protocols we will support, and what form the temporary resource will take.  i.e. do we do what scikit image does and download the file completely before dispatching to plugins? (that seems like it could cause problems in big data cases).  do we somehow try to mock a remote resource enough to know whether a plugin "could handle it" if we grabbed it for them? etc...  Do we try to be a little magical depending on the number of bytes?  Do we let the user decide with a "download this?" dialog

cc @perlman, @jni, @manzt, @sofroniewn

**Comment by perlman** [comment]
Are you looking to create a new hookspec for data retrieval? ;-)

I could see a plugin choosing one of three modes: fully wrapped (local file, default), stream of bytes, and path/url.
[fsspec](https://github.com/intake/filesystem_spec) may fit in here as well.

**Comment by sofroniewn** [comment]
@nclack and i were discussing the "stream of bytes" idea in the context of image acquisition tools the other day too - good to think about that use case more at some point too (though i don't want to hijak this thread!)

**Comment by perlman** [comment]
@sofroniewn Yeah, I was thinking more of file formats that can be read as a stream (like .npz).

**Comment by tlambert03** [comment]
+1 on fsspec.  

Was kind of thinking this would be a core feature, more than a hookspec, but everything is on the table!

I agree the stream of bytes has parallels.  Good to keep in mind (even if they diverge ultimately)

**Comment by neuromusic** [comment]
for "medium" data (too big to fetch every time, but small enough that you don't need zarr or streaming), pooch may come in handy https://github.com/fatiando/pooch
